lxd/cluster/gateway.go

```diff
@@ -177,25 +177,28 @@ func (g *Gateway) HandlerFuncs(nodeRefreshTask func(*APIHeartbeat)) map[string]h
 				return
 			}
 
-			nodes := make([]db.RaftNode, 0)
-			for _, node := range heartbeatData.Members {
-				if node.Raft {
-					nodes = append(nodes, db.RaftNode{
-						ID:      node.ID,
+			raftNodes := make([]db.RaftNode, 0)
+			for _, node := range heartbeatData.Nodes {
+				if node.RaftID > 0 {
+					raftNodes = append(raftNodes, db.RaftNode{
+						ID:      node.RaftID,
 						Address: node.Address,
 					})
 				}
 			}
 
-			// Accept Raft node updates from any node (joining nodes just send raft nodes heartbeat data).
-			logger.Debugf("Replace current raft nodes with %+v", nodes)
-			err = g.db.Transaction(func(tx *db.NodeTx) error {
-				return tx.RaftNodesReplace(nodes)
-			})
-			if err != nil {
-				logger.Errorf("Error updating raft nodes: %v", err)
-				http.Error(w, "500 failed to update raft nodes", http.StatusInternalServerError)
-				return
+			// Check we have been sent at least 1 raft node before wiping our set.
```

Comment:
Under which circumstances do we expect to have 0 raft nodes? A leader that is running an older version which does not fill the Nodes field?

Reply:
@freeekanayaka I added this as a safety check in the assumption that a heartbeat from any node should never contain 0 raft nodes. Is this a correct assumption? I added it mainly in the case that an old leader sends a heartbeat in the old format without the `Nodes` field (in which case there would be zero received after unmarshalling).

Reply:
It'd be nice to check if in that case the slice is set to ```nil``` (instead of a slice with zero elements). I believe that should be the case. So checking for nil would be a more explicit indication, and we could blow up if we receive a non-nil slice with 0 elements. Again, YMMV, just preference.

Reply:
@freeekanayaka sounds good, I'll check whether Go initialises this or not and if not will add a nil check too.

Reply:
@freeekanayaka added check and error log.